dask_jobqueue/core.py

```diff
@@ -14,17 +19,116 @@ class JobQueueCluster(object):
+    cancel_command: str
+        Abstract attribute for job scheduler cancel command, should be overriden
+
     See Also
     --------
     PBSCluster
     SLURMCluster
     """
-    def __init__(self):
-        raise NotImplemented
+
+    _script_template = """
+#!/bin/bash
+
+%(job_header)s
+
+%(worker_command)s
+""".lstrip()
+
+    #Following class attributes should be overriden by extending classes.
+    submit_command = None
+    cancel_command = None
+
+    def __init__(self,
+                 name='dask-worker',
+                 threads=4,
+                 processes=6,
+                 memory='16GB',
+                 interface=None,
+                 death_timeout=60,
+                 local_directory=None,
+                 extra='',
+                 **kwargs
+                 ):
+        """
+        This initializer should be considered as Abstract, and never used directly.
+        """
+        if not self.cancel_command or not self.submit_command:
+            raise NotImplementedError('JobQueueCluster is an abstract class that should not be instanciated.')
+
+        if interface:
+            host = get_ip_interface(interface)
+            extra += ' --interface  %s ' % interface
+        else:
+            host = socket.gethostname()
+
+        self.cluster = LocalCluster(n_workers=0, ip=host, **kwargs)
+
+        self.jobs = dict()
+        self.n = 0
+        self._adaptive = None
+
+        #dask-worker command line build
+        self._command_template = "%s/dask-worker %s" % (dirname, self.scheduler.address)
+        if threads is not None:
+            self._command_template += " --nthreads %d" % threads
+        if processes is not None:
+            self._command_template += " --nprocs %d" % processes
+        if memory is not None:
+            self._command_template += " --memory-limit %s" % memory
+        if name is not None:
+            self._command_template += " --name %s" % name
+            self._command_template += "-%(n)d" #Keep %(n) to be replaced later.
+        if death_timeout is not None:
+            self._command_template += " --death-timeout %s" % death_timeout
+        if local_directory is not None:
+            self._command_template += " --local-directory %s" % local_directory
+        if extra is not None:
+            self._command_template += extra
 
     def job_script(self):
         self.n += 1
-        return self._template % toolz.merge(self.config, {'n': self.n})
+        return self._script_template % {'job_header': self.job_header,
+                                        'worker_command': self._command_template % {'n': self.n}
+                                        }
+
+    @property
+    def job_header(self):
+        """
+        Abstract attribute for the Job scheduler script header part.
+
+        Returns
+        -------
+        A string containing multiple lines to be used as header of job file to be sumitted.
+        """
+        raise NotImplementedError
```

Comment:
Let's just make this a class attribute that defaults to an empty string.

Reply:
Okay, I tried to keep one of my "contract" of Abstract Class, but I guess it is still too Java oriented :).